engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -294,19 +297,44 @@ - (void)sharedSetupWithProject:(nullable FlutterDartProject*)project
                                                 opaque:_viewOpaque
                                        enableWideGamut:engine.project.isWideGamutEnabled];
   [_engine createShell:nil libraryURI:nil initialRoute:initialRoute];
+  [self performImplicitEngineCallbacks];  // This is the earliest this can be called because it
+                                          // depends on the shell being created.
   _engineNeedsLaunch = YES;
   _ongoingTouches = [[NSMutableSet alloc] init];
 
   // TODO(cbracken): https://github.com/flutter/flutter/issues/157140
   // Eliminate method calls in initializers and dealloc.
   [self loadDefaultSplashScreenView];
   [self performCommonViewControllerInitialization];
+}
+
+- (void)performImplicitEngineCallbacks {
+  // We call this from the FlutterViewController instead of the FlutterEngine directly because this
+  // is only needed when the FlutterEngine is implicit. If it's not implicit there's no need for
+  // them to have a callback to expose the engine since they created the FlutterEngine directly.
+  BOOL notified = [_engine performAppDelegateEngineInitializationCallback];
 
+  // TODO(vashworth): Deprecate, see https://github.com/flutter/flutter/issues/176424
   if ([FlutterSharedApplication.application.delegate
           respondsToSelector:@selector(pluginRegistrant)]) {
     NSObject<FlutterPluginRegistrant>* pluginRegistrant =
         [FlutterSharedApplication.application.delegate performSelector:@selector(pluginRegistrant)];
     [pluginRegistrant registerWithRegistry:self];
```

Comment:
I know this is existing code, but the context and my comment above highlighted it for me, and it's *really* dangerous since it's giving out external `self` reference before initialization is complete. Can we just change this to `registerWithRegistry:engine`?

(If you want to punt on this that's fine, but since it's related and I think it's probably an easy fix I wanted to flag it.)

Reply:
Having thought about this more, we shouldn't change this here given the timing; if it causes some subtle problem we don't want that to impact getting this stuff landed. I'll make a note to look at this soon myself.